### PR TITLE
reflection sync

### PR DESCRIPTION
**Pull request by xeno-by**
review @retronym @adriaanm

also see https://docs.google.com/document/d/1GmHRmqahh1tTH_GOPsRlnGVRAFPe-x0iCu3BNMNXjf0/edit

### CODE REVIEW

**Comment by retronym** [comment]
The IDE team keep an eye on those and reported that it was spurious.

```
Tests in error: 
  findTestClass_WhenThereAreParseErrors(scala.tools.eclipse.launching.JUnit4TestFinderTest): /runAsJunit/src/test/MyTest.scala already exists in target
  findSimpleTestClass_thatUsesRunWithAnnotation(scala.tools.eclipse.launching.JUnit4TestFinderTest): /runAsJunit/src/test/MyTest.scala already exists in target
```

They identified the problem and have merged a fix to their test suite: https://github.com/scala-ide/scala-ide/pull/566

**Comment by retronym** [comment]
PLS REBUILD/pr-scala@7fec1e5
PLS REBUILD/pr-scala@02c013d
PLS REBUILD/pr-scala@1b223bd
PLS REBUILD/pr-scala@2d66925
PLS REBUILD/pr-scala@3b02963
PLS REBUILD/pr-scala@55fe060
PLS REBUILD/pr-scala@281e7a6
PLS REBUILD/pr-scala@62cf091
PLS REBUILD/pr-scala@73c3ee1
PLS REBUILD/pr-scala@2ca0980
PLS REBUILD/pr-scala@11e6b57
PLS REBUILD/pr-scala@7206363
PLS REBUILD/pr-scala@0aabd05
PLS REBUILD/pr-scala@7ded6c8
PLS REBUILD/pr-scala@e0f4dd5
PLS REBUILD/pr-scala@cc16292
PLS REBUILD/pr-scala@299a7ab

**Comment by retronym** [comment]
NOLITTER!

**Comment by xeno-by** [comment]
Fixed non-spurious build errors and rebased

**Comment by retronym** [comment]
PLS REBUILD/pr-scala@e2466b2

(@xeno-by I've just configured Jenkins pr-scala-test job to archive `scala/build/junit/TEST-*,scala/build/osgi/TEST-*` so the JUnit test failures will be visible as a build artifact.)

**Comment by xeno-by** [comment]
@retronym Much appreciated! That's what I wanted to do myself this afternoon. How did you do that?

**Comment by xeno-by** [comment]
@retronym It works! Amazing!! https://scala-webapps.epfl.ch/jenkins/view/All/job/pr-scala-test/2940/artifact/scala/build/osgi/TEST-tools.test.osgi.reflection.toolbox.ReflectionToolBoxTest.xml

**Comment by retronym** [comment]
Add post build step; archive artifacts.

File: src/reflect/scala/reflect/runtime/JavaUniverse.scala [hunk 1]
```diff
@@ -27,4 +27,73 @@ class JavaUniverse extends internal.SymbolTable with ReflectSetup with runtime.S
...
+  // Before reading the text below, read up the section Mirrors in the reflection pre-SIP
+  // https://docs.google.com/document/d/1nAwSw4TmMplsIlzh2shYLUJ5mVh3wndDa1Zm1H6an9A/edit.
+  // Take an especially good look at Figure 2, because it illustrates fundamental principles underlying runtime reflection:
+  //   1) For each universe we have one mirror per classloader
+  //   2) Package symbols are per-mirror
+  //   3) Other symbols are per-universe, which means that a symbol (e.g. Seq on the picture) might be shared between multiple owners
+  //
+  // Main challenges that runtime reflection presents wrt initialization are:
+  //   1) Extravagant completion scheme that enters package members on-demand rather than a result of scanning a directory with class files.
+  //      (That's a direct consequence of the fact that in general case we can't enumerate all classes in a classloader).
```
Comment (paulp) [inline src/reflect/scala/reflect/runtime/JavaUniverse.scala]: It isn't. It's a direct consequence of the fact that we don't do it.

The fact is that in almost every case - certainly everything I ever do, and I don't think I'm unusual - all the classes are known at startup. That this is not ALWAYS the case is not a reason to utilize an "extravagant completion scheme" 100% of the time.

File: test/files/run/reflection-sync-subtypes.check [hunk 2]
```diff
@@ -0,0 +1,1000 @@
+false, false, true, true
```
Comment (paulp) [inline test/files/run/reflection-sync-subtypes.check]: When a checkfile reveals interesting attributes of the behaviors being tested, it is nice to see the output. This is not interesting. The "println" line would be better as an "assert" and lose this entire file.

File: src/compiler/scala/tools/reflect/ToolBoxFactory.scala [hunk 3]
```diff
@@ -340,7 +340,8 @@ abstract class ToolBoxFactory[U <: JavaUniverse](val u: U) { factorySelf =>
         lazy val exporter = importer.reverse
       }
 
-      def apply[T](f: CompilerApi => T): T = {
+      private val toolBoxLock = new Object
+      def apply[T](f: CompilerApi => T): T = toolBoxLock.synchronized {
```
Comment (retronym) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala:344]: > 2 additions 1 deletion

Hey, our little `withCompilerApi` turned out nicely :)

File: test/files/run/reflection-sync-potpourri.scala [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+  def force(lazytpe: () => Type): String = {
+    lazytpe().typeSymbol.typeSignature
+    lazytpe().toString
+  }
+  val diceRolls = List.fill(n)(rng.nextInt(perms.length))
+  val threads = (1 to n) map (i => new Thread(s"Reflector-$i") {
+    override def run(): Unit = {
+      val s1 = foo("42")
+      val s2 = perms(diceRolls(i - 1)).map(x => force(x)).sorted.mkString(", ")
+      println(s"s1 = $s1, s2 = $s2")
```
Comment (retronym) [inline test/files/run/reflection-sync-potpourri.scala]: In the spirit of Paul's other review comment, we should probably change this to an assert and lose the repetitive check file.

**Comment by retronym** [comment]
f730ee6f40d560b2fc4f00047668f3909c94e9c1 ("tests for fancy classloader configurations") looks like it was the result of some painstaking forensic work. Could you share some of that in code / commit comments?

File: src/compiler/scala/tools/reflect/ToolBoxFactory.scala (hunk 3 above)
Comment (xeno-by) [inline src/compiler/scala/tools/reflect/ToolBoxFactory.scala:344]: :)

**Comment by retronym** [comment]
@xeno-by ping for two review comments (remove meaninless check files / document classloader magic test.)

**Comment by xeno-by** [comment]
Will do. Is everything else lgtm?

**Comment by retronym** [comment]
Yep.

File: test/files/run/reflection-sync-subtypes.check (hunk 2 above)
Comment (xeno-by) [inline test/files/run/reflection-sync-subtypes.check]: Done

File: test/files/run/reflection-sync-potpourri.scala (hunk 4 above)
Comment (xeno-by) [inline test/files/run/reflection-sync-potpourri.scala]: Done

**Comment by xeno-by** [comment]
@retronym re f730ee6: The configurations being tested aren't anything special and aren't a consequence of me obtaining some hidden knowledge. I just wanted to bring a bit more chaos to our testing, and these two tests represent some of the possibilities to achieve that.

File: src/reflect/scala/reflect/runtime/JavaUniverse.scala (hunk 1 above)
Comment (xeno-by) [inline src/reflect/scala/reflect/runtime/JavaUniverse.scala]: Updated the comment

**Comment by xeno-by** [comment]
Everything looks good:
- The commits that we went through together with Jason were lgtm'd and documented in https://docs.google.com/document/d/1GmHRmqahh1tTH_GOPsRlnGVRAFPe-x0iCu3BNMNXjf0/edit
- The pull request feedback is addressed
- All commits except for 089cbc7 have green checkmarks next to them
- 089cbc7, still yellow on github, has its job recently turned green (https://scala-webapps.epfl.ch/jenkins/job/pr-scala/3905/)

Since the pull request is huge, I'll be merging right now without waiting for the github bot to pick up the last remaing job status in order to avoid potential conflicts that would lead to more rebases.

**Comment by paulp** [comment]
@xeno-by My scala doesn't work anymore. Do you know what's going on?

```
/scala/inst/scala-2.11.0-20131018-163133-e3666d5bf5/bin/scala
Welcome to Scala version 2.11.0-20131018-163133-e3666d5bf5 (Java HotSpot(TM) 64-Bit Server VM, Java 1.7.0_40).
Type in expressions to have them evaluated.
Type :help for more information.

scala> 123
java.lang.AssertionError: assertion failed: pkgClass = package lang, sym = class $1, existing = class $1 (depth=0)
    at scala.Predef$.assert(Predef.scala:169)
    at scala.reflect.runtime.SymbolLoaders$PackageScope.enter(SymbolLoaders.scala:100)
    at scala.reflect.runtime.SymbolLoaders$class.initAndEnterClassAndModule(SymbolLoaders.scala:45)
    at scala.reflect.runtime.JavaUniverse.initAndEnterClassAndModule(JavaUniverse.scala:11)
    at scala.reflect.runtime.JavaMirrors$JavaMirror.scala$reflect$runtime$JavaMirrors$JavaMirror$$jclassAsScala1(JavaMirrors.scala:1069)
```

b43f13c741 is fine.

**Comment by paulp** [comment]
Looks like it's java7 vs. java6 - java6 works. Is java7 still completely untested, how is that possible?

**Comment by retronym** [comment]
@xeno-by The dose of chaos in `t6240{a,b}` seems to chaotic for `scala-checkin-manual-windows`: https://scala-webapps.epfl.ch/jenkins/job/scala-checkin-manual-windows/27/. The assertion for `File(javaBinary).exists` fails. I don't know why...

**Comment by xeno-by** [comment]
@retronym Will take a look.

**Comment by xeno-by** [comment]
Oh man, it should be `C:\Java\jdk1.6.0_24-x64\jre\bin\java.exe` not `C:\Java\jdk1.6.0_24-x64\jre\bin\java`. I guess I'll just remove the assert.

**Comment by xeno-by** [comment]
@retronym https://github.com/scala/scala/pull/3062

**Comment by adriaanm** [comment]
https://scala-webapps.epfl.ch/jenkins/job/pr-scala-test/3142/consoleText looks like an example of a spurious failure -- from a job based on current master:

```
% diff /localhome/jenkins/a/workspace/pr-scala-test/scala/test/files/run/reflection-sync-subtypes-run.log /localhome/jenkins/a/workspace/pr-scala-test/scala/test/files/run/reflection-sync-subtypes.check
@@ -1,100 +0,0 @@
-Exception in thread "Reflector-774" scala.reflect.internal.MissingRequirementError: term tasks  in Test not found.
-   at scala.reflect.internal.MissingRequirementError$.signal(MissingRequirementError.scala:17)
-   at scala.reflect.internal.MissingRequirementError$.notFound(MissingRequirementError.scala:18)
-   at scala.reflect.internal.BuildUtils$BuildImpl.select(BuildUtils.scala:26)
-   at scala.reflect.internal.BuildUtils$BuildImpl.selectTerm(BuildUtils.scala:17)
-   at scala.reflect.internal.BuildUtils$BuildImpl.selectTerm(BuildUtils.scala:11)
-   at Test$$anonfun$2$$typecreator3$1.apply(reflection-sync-subtypes.scala:8)
-   at scala.reflect.api.TypeTags$WeakTypeTagImpl.tpe$lzycompute(TypeTags.scala:231)
-   at scala.reflect.api.TypeTags$WeakTypeTagImpl.tpe(TypeTags.scala:231)
-   at scala.reflect.api.TypeTags$class.typeOf(TypeTags.scala:335)
-   at scala.reflect.api.Universe.typeOf(Universe.scala:60)
-   at Test$$anonfun$2.apply$mcZ$sp(reflection-sync-subtypes.scala:8)
-   at Test$$anonfun$6$$anon$1$$anonfun$7.apply(reflection-sync-subtypes.scala:15)
-   at Test$$anonfun$6$$anon$1$$anonfun$7.apply(reflection-sync-subtypes.scala:15)
-   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
-   at scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:245)
-   at scala.collection.immutable.List.foreach(List.scala:302)
-   at scala.collection.TraversableLike$class.map(TraversableLike.scala:245)
-   at scala.collection.AbstractTraversable.map(Traversable.scala:104)
-   at Test$$anonfun$6$$anon$1.run(reflection-sync-subtypes.scala:15)
```

**Comment by xeno-by** [comment]
Thanks! I'll be looking into this today or tomorrow. We discussed this with Jason, and he suggested that I bring down the tests at fault in order to block people's work and then calmly figure out what's wrong with them. I'll submit a pull request for that shortly.

**Comment by xeno-by** [comment]
@adriaanm https://github.com/scala/scala/pull/3071

**Pull request merged**